File: Topics/BTree/BTLevelOrderTraversal.java
package Topics.BTree;

import Libs.TreeNode;

import java.util.*;

public class BTLevelOrderTraversal {
    public List<List<Integer>> levelOrder(TreeNode root) {
//        List<List<Integer>> re = new ArrayList<>();
//        if (root == null) {
//            return re;
//        }
//        Queue<TreeNode> q = new LinkedList<>();
//        q.add(root);
//        while (!q.isEmpty()) {
//            int size = q.size();
//            re.add(new ArrayList<>());
//            for (int i = 0; i < size; i++) {
//                TreeNode cur = q.poll();
//                re.get(re.size()-1).add(cur.val);
//                if (cur.left != null) {
//                    q.add(cur.left);
//                }
//                if (cur.right != null) {
//                    q.add(cur.right);
//                }
//            }
//        }
//        return re;

        List<List<Integer>> re = new ArrayList<>();
        helper(re, root, 0);
        return re;
    }

    private void helper(List<List<Integer>> re, TreeNode root, int level) {
        if (root != null) {
            if (level == re.size()) {
                re.add(new ArrayList<>());
            }
            re.get(level).add(root.val);
            helper(re, root.left, level+1);
            helper(re, root.right, level+1);
        }
    }
}

File: Topics/BTree/BTLevelOrderTraversal.cpp
class Solution { // BFS
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> result;
        if (!root) {
            return result;
        }

        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            int size = q.size();
            result.emplace_back();

            for (int i = 0; i < size; ++i) {
                TreeNode* cur = q.front();
                q.pop();

                result.back().push_back(cur->val);

                if (cur->left) {
                    q.push(cur->left);
                }
                if (cur->right) {
                    q.push(cur->right);
                }
            }
        }

        return result;
    }
};

class Solution { // DFS
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> result;
        dfs(root, 0, result);
        return result;
    }

private:
    void dfs(TreeNode* node, int level, vector<vector<int>>& result) {
        if (!node) {
            return;
        }

        if (level == result.size()) {
            result.emplace_back();
        }

        result[level].push_back(node->val);

        dfs(node->left, level + 1, result);
        dfs(node->right, level + 1, result);
    }
};
